File: Python/ByTier/F/1704_F_Colouring_Game.py
# Problem: CF 1704 F - Colouring Game
# https://codeforces.com/contest/1704/problem/F

"""
Algorithm: Game Theory with Sprague-Grundy Theorem
This problem is a classic impartial game that can be solved using the Sprague-Grundy theorem.
We compute the Grundy number (nimber) for each segment of consecutive same-colored cells.
The overall Grundy number is the XOR of all segment Grundy numbers.
If the total Grundy number is 0, the current player (Alice) loses; otherwise, Alice wins.

Time Complexity: O(n) per test case
Space Complexity: O(1) for precomputed Grundy numbers, O(n) for input processing
"""

f = [0] * int(6e5)  # Precompute Grundy numbers for all possible segment lengths
for i in range(1, 2000):  # Compute Grundy numbers for small values using mex
    seen = set()
    for j in range(0, i - 1):  # For each possible split in segment of length i
        seen.add(f[j] ^ f[i - 2 - j])  # XOR Grundy numbers of left and right parts
    mex = 0  # Find the minimum excludant (mex)
    while mex in seen:
        mex += 1
    f[i] = mex  # Store computed Grundy number

# For large values, use periodicity (after 2000, sequence repeats every 34)
for i in range(2000, int(6e5)):
    f[i] = f[i - 34]

t = int(input())  # Read number of test cases
for _ in range(t):
    n = int(input())  # Read length of string
    s = input()  # Read the string of colors
    
    # Count net difference between red and blue cells
    diff = s.count("R") - s.count("B")
    
    # If reds dominate, Alice wins immediately
    if diff > 0:
        print("Alice")
    # If blues dominate, Bob wins immediately
    elif diff < 0:
        print("Bob")
    else:
        # Otherwise, analyze the segments of same-colored cells
        ans = 0  # Initialize the overall XOR of segments
        i = 0
        lst = []
        while i < n:
            j = i + 1
            # Extend j until color changes
            while j < n and s[j] != s[j - 1]:
                j += 1
            lst.append(j - i)  # Store length of this segment
            i = j  # Move to start of next segment
            
        # Compute XOR of Grundy numbers for each segment
        for i in lst:
            ans ^= f[i]
            
        # If final XOR is 0, Bob wins; otherwise Alice wins
        print("Alice" if ans else "Bob")


# https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/1704/1704_F_Colouring_Game.cpp
// Problem: CF 1704 F - Colouring Game
// https://codeforces.com/contest/1704/problem/F

/*
Algorithm/Techniques: 
This problem is a classic game theory problem involving nim-like gameplay with specific rules.
The approach uses:
1. **Nim Game Theory**: The game can be broken down into independent segments (groups of same-colored cells), and each group contributes to the overall XOR (nim-sum) of the game.
2. **Precomputed SG Values**: An array `sg` is used, which stores the Sprague-Grundy values for different group lengths. These values help determine the outcome of the game.
3. **Segmentation**: The string is segmented by color changes into groups of consecutive red or blue cells.
4. **Special Case Handling**: Based on the net count of red vs blue cells (`d`), the winner is decided directly if `d != 0`.

Time Complexity: O(n) per test case due to linear traversal of the string and constant-time operations for SG lookup.
Space Complexity: O(1) since the precomputed `sg` array is fixed and does not depend on input size; only a constant amount of extra space is used.
*/

#include <stdio.h>
#include <iostream>

using namespace std;

// Precomputed Sprague-Grundy values for group lengths
const int sg[87] = {0, 0, 1, 1, 2, 0, 3, 1, 1, 0, 3, 3, 2, 2, 4, 0, 5, 2,
                    2, 3, 3, 0, 1, 1, 3, 0, 2, 1, 1, 0, 4, 5, 2, 7, 4, 0,
                    1, 1, 2, 0, 3, 1, 1, 0, 3, 3, 2, 2, 4, 4, 5, 5, 2, 3,
                    3, 0, 1, 1, 3, 0, 2, 1, 1, 0, 4, 5, 3, 7, 4, 8, 1, 1,
                    2, 0, 3, 1, 1, 0, 3, 3, 2, 2, 4, 4, 5, 5, 9};

int t, n, d, ans;
char s[500005];

int main() {
  ios::sync_with_stdio(0), cin.tie(0), cin >> t;
  while (t--) {
    // Read input string and set boundary condition
    cin >> n >> s + 1, s[n + 1] = s[n];
    
    // Count net red - blue cells
    d = 0;
    for (int i = 1; i <= n; i++)
      s[i] == 'R' ? d++ : d--;
    
    // Process segments of same-colored cells to compute nim-sum
    ans = 0;
    for (int i = 1, cnt = 0; i <= n + 1; i++) {
      if (s[i] == s[i - 1])
        // XOR with Sprague-Grundy value for current segment length
        ans ^= sg[cnt <= 52 ? cnt : (cnt - 53) % 34 + 53], cnt = 1;
      else
        cnt++;
    }
    
    // Determine winner based on net balance and nim-sum
    puts(d ? d > 0 ? "Alice" : "Bob" : ans ? "Alice" : "Bob"), d = ans = 0;
  }
  return 0;
}


// https://github.com/VaHiX/CodeForces/